File: src/QuranCompanion.Api/Common/ApiResponse.cs
namespace QuranCompanion.Api.Common;

public sealed record ApiResponse<T>(
    T? Data,
    bool Success,
    string? Message,
    IReadOnlyList<ApiError> Errors)
{
    public static ApiResponse<T> Ok(T data) =>
        new(data, true, null, Array.Empty<ApiError>());

    public static ApiResponse<T> Fail(string message, IReadOnlyList<ApiError>? errors = null) =>
        new(default, false, message, errors ?? Array.Empty<ApiError>());
}

public sealed record ApiError(string Code, string Message, string? Field = null);

File: apps/ios/QuranMeali/Api/ApiResponse.swift
//
//  ApiResponse.swift
//  QuranMeali
//
//  Envelope returned by every QuranCompanion endpoint.
//  { "data": T?, "success": bool, "message": string?, "errors": ApiError[] }
//

import Foundation

struct ApiResponse<T: Decodable>: Decodable {
    let data: T?
    let success: Bool
    let message: String?
    let errors: [ApiResponseError]?
}

struct ApiResponseError: Decodable {
    let code: String
    let message: String
    let field: String?
}
